build-system-abstraction.rst

```diff
@@ -0,0 +1,456 @@
+setup-requirement by a second project (B) which has not itself transitioned to
+having a pypa.json will make B uninstallable by any version of pip. This is
+because setup.py in B will trigger easy-install when 'setup.py egg_info' is
+run by pip, and that will try and fail to install A.
+
+As such we recommend that tools which are currently used as setup-requires
+either ensure that they keep a `setuptools shim`_ or find their consumers and
+get them all to upgrade to the use of a `pypa.json` in advance of moving
+themselves. Pragmatically that is impossible, so the advice is to keep a
+setuptools shim indefinitely - both for projects like pbr, setuptools_scm and
+also projects like numpy.
+
+setuptools shim
+---------------
+
+It would be possible to write a generic setuptools shim that looks like
+``setup.py`` and under the hood uses ``pypa.json`` to drive the builds. This
+is not needed for pip to use the system, but would allow package authors to
+use the new features while still retaining compatibility with older pip
+versions.
+
+Rationale
+=========
+
+This PEP started with a long mailing list thread on distutils-sig [#thread]_.
+Subsequent to that a online meeting was held to debug all the positions folk
+had.  Minutes from that were posted to the list [#minutes]_.
+
+This specification is a translation of the consensus reached there into PEP
+form, along with some arbitrary choices on the minor remaining questions.
+
+The basic heuristic for the design has been to focus on introducing an
+abstraction without requiring development not strictly tied to the
+abstraction. Where the gap is small to improvements, or the cost of using the
+existing interface is very high, then we've taken on having the improvement as
+a dependency, but otherwise defered such to future iterations.
+
+We chose wheel METADATA files rather than defining a new specification,
+because pip can already handle wheel .dist-info directories which encode all
+the necessary data in a METADATA file. PEP-426 can't be used as it's still
+draft, and defining a new metadata format, while we should do that, is a
+separate problem. Using a directory on disk would not add any value to the
+interface (pip has to do that today due to limitations in the setuptools
+CLI).
+
+The use of 'develop' as a command is because there is no PEP specifying the
+interoperability of things that do what 'setuptools develop' does - so we'll
+need to define that before pip can take on the responsibility for doing the
+'develop' step. Once that's done we can issue a successor PEP to this one.
+
+The use of a command line API rather than a Python API is a little
+contentious. Fundamentally anything can be made to work, and the pip
+maintainers have spoken strongly in favour of retaining a process based
+interface - something that is mature and robust in pip today.
+
+The choice of JSON as a file format is a compromise between several
```

Comment:
It does, but it's a much simpler format, and the primary module for it is pure python, whereas PyYAML tries to build an extension module.

Personally, I do like toml as a format - I'm considering transitioning to toml config files for flit. But these metadata files, as currently proposed, are so small and rarely changing that I don't think it's a problem to work with them in JSON.

If we do consider `pypa.toml`, though, I'd like to suggest a namespace which build tools can use to store their own metadata, to reduce the clutter of separate files a bit:

```toml
schema = 1
bootstrap_requires = ["flit"]
build_command = ["flit"]

[build.flit]
module = "foobar"
# ...
```

The build system would be responsible for reading the file again if it wanted to use this, and there would be no requirement for build tools to use this file at all.

This would be more complexity than I'd want to put in a JSON file, though.